python/cudf/cudf/core/frame.py

```diff
@@ -34,6 +37,135 @@ class Frame(libcudfxx.table.Table):
     def _from_table(cls, table):
         return cls(table._data, index=table._index)
 
+    @classmethod
+    def _concat(cls, objs, axis=0, ignore_index=False):
+
+        # shallow-copy the input DFs in case the same DF instance
+        # is concatenated with itself
+        objs = [df.copy(deep=False) for df in objs]
+
+        # TODO (ptaylor):
+        # libcudf.nvtx.nvtx_range_push("CUDF_CONCAT", "orange")
+
+        from cudf.core.column.column import column_empty
+        from cudf.core.column.column import build_categorical_column
+
+        if ignore_index:
+            index = cudf.core.index.RangeIndex(sum(map(len, objs)))
+        elif isinstance(objs[0].index, cudf.core.multiindex.MultiIndex):
+            index = cudf.core.multiindex.MultiIndex._concat(
+                [o.index for o in objs]
+            )
+        else:
+            index = cudf.core.index.Index._concat([o.index for o in objs])
```

Comment:
I left this in here because I'm not sure whether/how `concatenate` should handle MultiIndex at the cython/libcudf layer. Is libcudf expected to handle MultiIndexes? Should this branching happen in the Cython instead and use `table.view()` if normal index, `tbl.data_view()` if MultiIndex?

Reply:
@trxcllnt if you call `.view()` it just has all of the index columns (whether Index or MultiIndex) as the first columns of the table. Then when passing in `index_names` in the `from_unique_ptr` call it reads those from the left most columns first. You should be able to just call `.view()` and pass the index names appropriate and everything will just work in a single libcudf API call.